src/GafferScene/ShufflePrimitiveVariables.cpp

```diff
@@ -0,0 +1,133 @@
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "GafferScene/ShufflePrimitiveVariables.h"
+
+#include "GafferScene/SceneAlgo.h"
+
+#include "IECoreScene/Primitive.h"
+
+using namespace std;
+using namespace IECore;
+using namespace IECoreScene;
+using namespace Gaffer;
+using namespace GafferScene;
+
+GAFFER_GRAPHCOMPONENT_DEFINE_TYPE( ShufflePrimitiveVariables );
+
+size_t ShufflePrimitiveVariables::g_firstPlugIndex = 0;
+
+ShufflePrimitiveVariables::ShufflePrimitiveVariables( const std::string &name ) : Deformer( name )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+	addChild( new ShufflesPlug( "shuffles" ) );
+}
+
+ShufflePrimitiveVariables::~ShufflePrimitiveVariables()
+{
+}
+
+Gaffer::ShufflesPlug *ShufflePrimitiveVariables::shufflesPlug()
+{
+	return getChild<Gaffer::ShufflesPlug>( g_firstPlugIndex );
+}
+
+const Gaffer::ShufflesPlug *ShufflePrimitiveVariables::shufflesPlug() const
+{
+	return getChild<Gaffer::ShufflesPlug>( g_firstPlugIndex );
+}
+
+bool ShufflePrimitiveVariables::affectsProcessedObject( const Gaffer::Plug *input ) const
+{
+	return Deformer::affectsProcessedObject( input ) || shufflesPlug()->isAncestorOf( input );
+}
+
+void ShufflePrimitiveVariables::hashProcessedObject( const ScenePath &path, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	Deformer::hashProcessedObject( path, context, h );
+
+	shufflesPlug()->hash( h );
+}
+
+IECore::ConstObjectPtr ShufflePrimitiveVariables::computeProcessedObject( const ScenePath &path, const Gaffer::Context *context, const IECore::Object *inputObject ) const
+{
+	auto inputPrimitive = runTimeCast<const Primitive>( inputObject );
+	if( !inputPrimitive )
+	{
+		return inputObject;
+	}
+
+	if( shufflesPlug()->children().empty() )
+	{
+		return inputObject;
+	}
+
+	PrimitivePtr result = inputPrimitive->copy();
+	shufflesPlug()->shuffle<PrimitiveVariableMap>( inputPrimitive->variables, result->variables );
+
+	return result;
+}
+
+bool ShufflePrimitiveVariables::adjustBounds() const
+{
+	if( !Deformer::adjustBounds() )
+	{
+		return false;
+	}
+
+	for( auto &child : shufflesPlug()->children() )
+	{
+		ShufflePlug *shuffle = static_cast<ShufflePlug *>( child.get() );
+
+		if( shuffle->enabledPlug()->getValue() )
+		{
+			std::string destination = shuffle->destinationPlug()->getValue();
+			if( destination == "P" )
+			{
+				return true;
+			}
```

Comment:
Technically we should be providing `${source}` here, but I'm reluctant to make this any more complicated given that it's unlikely to matter in practice, and we'll probably refactor into a Shuffler later. Perhaps just a comment?

Reply:
I could do it without scoping a context, right? `dest == "P" && ( source == "$source" || source == "${source}"`

Reply:
Oh wait, I did that below... ok, so I'll ditch this then?

Reply:
Addressed in e810578